### PR TITLE
Dev bcewithlogits loss

### PR DESCRIPTION
**Pull request by MARD1NO**
1. 增加 BCEWithLogitsLoss 算子
2. 去除了 原始 BCELoss 引入的sigmoid 运算,但是缺乏对数值范围的检查,需要保证数值在 0-1, 后续到C++后端重写算子

BCEWithLogitsLoss 相当于 sigmoid + bceloss, 但是Pytorch为了更好的数值稳定性, 没有直接使用sigmoid运算, 而是做了一些数学操作, 下面是其源代码

```c++
Tensor binary_cross_entropy_with_logits(const Tensor& input, const Tensor& target, const Tensor& weight, const Tensor& pos_weight, int64_t reduction) {
    Tensor loss;
    auto max_val = (-input).clamp_min_(0);
    if (pos_weight.defined()) {
        // pos_weight need to be broadcasted, thus mul(target) is not inplace.
        auto log_weight = (pos_weight - 1).mul(target).add_(1);
        loss = (1 - target).mul_(input).add_(log_weight.mul_(((-max_val).exp_().add_((-input - max_val).exp_())).log_().add_(max_val)));
    } else {
        loss = (1 - target).mul_(input).add_(max_val).add_((-max_val).exp_().add_((-input -max_val).exp_()).log_());
    }

    if (weight.defined()) {
        loss.mul_(weight);
    }

    return apply_loss_reduction(loss, reduction);
}
```
使用了大量数学操作,以及inplace操作

为了更清晰的表示整个计算思路,相关计算的python版本,会在单测中体现

### CODE REVIEW

File: oneflow/python/ops/nn_ops.py [hunk 1]
```diff
@@ -2786,19 +2786,19 @@ def bce_loss(
 
     .. math:: 
 
-        out = -(Target_i*ln(Input_i) + (1-Target_i)*ln(1-Input_i))
+        out = -(Target_i*log(Input_i) + (1-Target_i)*log(1-Input_i))
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:2789]: flow.math.log ，[底层实现](https://github.com/Oneflow-Inc/oneflow/blob/cd836a54364e6ad33ec7f6e84df597f56778ef7c/oneflow%2Fuser%2Fkernels%2Fmath_unary_elementwise_func.h#L225)是调用的 std::log，这个就是自然对数，文档的 ln 没错
https://en.cppreference.com/w/cpp/numeric/math/log

File: oneflow/python/test/ops/test_bce_with_logits_loss.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    flow.clear_default_session()
+    if device_type == "cpu":
+        flow.config.cpu_device_num(device_counts)
+    else:
+        flow.config.gpu_device_num(device_counts)
+
+    func_config = flow.FunctionConfig()
+    func_config.default_placement_scope(flow.scope.placement(device_type, machine_ids))
+
+    def np_bce_with_logits_loss(np_input, np_target, np_weight, np_pos_weight):
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_bce_with_logits_loss.py:52]: 貌似除了 np_weight 其他参数都没用到

File: oneflow/python/ops/nn_ops.py [hunk 3]
```diff
@@ -2877,6 +2877,138 @@ def bceloss_job(
...
+        "Found the length of pos_weight {} vs classes {}".format(
+            pos_weight.shape[0], input.shape[-1]
+        )
+    )
+
+    if name is None:
+        name = id_util.UniqueStr("BCEWithLogitsLoss")
+
+    _neg_input = flow.math.negative(input)
+    _max_val = flow.clip(_neg_input, min_value=0)
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:2978]: 这个操作怎么理解，对 输入取负数，然后 clip(0, xx)，也就是原来输入大于0的部分变成0，小于0的部分符号取反

File: oneflow/python/ops/nn_ops.py (hunk 1 above)
Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py:2789]: =，=我觉得很丑才这么改，各大框架的log其实都是ln，但是文档上面的公式都写的是 log  我很纠结

File: oneflow/python/ops/nn_ops.py (hunk 3 above)
Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py:2978]: =，= pytorch实现就是这样命名的，我又不太能理解他这样绕的数学原理是什么

File: oneflow/python/ops/nn_ops.py (hunk 1 above)
Comment (Ldpe2G) [reply oneflow/python/ops/nn_ops.py:2789]: 也行吧，可能大家都默认了 log 就是 ln

File: oneflow/python/ops/nn_ops.py (hunk 3 above)
Comment (Ldpe2G) [reply oneflow/python/ops/nn_ops.py:2978]: 好的

File: oneflow/python/ops/nn_ops.py [hunk 4]
```diff
@@ -2877,6 +2877,138 @@ def bceloss_job(
...
+    r"""This operator combines the `Sigmoid` and `BCELoss` together. For numerical stability, 
+    we apply some math tricks instead of using `Sigmoid` layer with `BCELoss`. 
+
+    The equation is: 
+
+    if reduction = "none": 
+
+    .. math:: 
+
+        out = -(Pos\_weight*Target_i*log(Input_i) + (1-Target_i)*log(1-Input_i))
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:2898]: 这里的公式和上面 bce_loss 的一样，没体现合并了 sigmoid ，可以参考一下 pytroch 的文档，把 input_i 用个 符号再包一下
image

**Comment by Ldpe2G** [comment]
推导了一下，知道 pytroch 的公式为什么是这个形式的了：

image-20201224150442549

下面是带 pos_weight 的版本：

image-20201224150456695



文档说是用到了 log-sum-exp 这个 trick，查了下资料
 https://www.xarg.org/2016/06/the-log-sum-exp-trick-in-machine-learning/

File: oneflow/python/ops/nn_ops.py [hunk 5]
```diff
@@ -2877,6 +2877,138 @@ def bceloss_job(
...
+
+    assert reduction in [
+        "none",
+        "mean",
+        "sum",
+    ], "{} is not a valid value for reduction, The reduction must be the one of `none`, `mean`, `sum`. ".format(
+        reduction
+    )
+
+    assert pos_weight.shape[0] == input.shape[-1], (
```
Comment (Ldpe2G) [inline oneflow/python/ops/nn_ops.py:2967]: input 的维度是要 >=2，加个检查吧

**Comment by MARD1NO** [comment]
> 推导了一下，知道 pytroch 的公式为什么是这个形式的了：
> 
> image-20201224150442549
> 
> 下面是带 pos_weight 的版本：
> 
> image-20201224150456695
> 
> 文档说是用到了 log-sum-exp 这个 trick，查了下资料
> https://www.xarg.org/2016/06/the-log-sum-exp-trick-in-machine-learning/

之前看过几个基于softmax的，现在好像更清晰了， 为了避免负数 在 计算 e^(-x) 出现溢出情况， 先反转，取到max，再间接通过

log-sum-exp 这个小技巧来避免了数值溢出

File: oneflow/python/test/ops/test_bce_with_logits_loss.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+def _compare_bce_with_logits_loss_np(
+    input_shape,
+    target_shape,
+    weight_shape,
+    pos_weight_shape,
+    device_type,
+    machine_ids,
+    device_counts,
+):
+    # random is clip to 0-1
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_bce_with_logits_loss.py:35]: 这句话有点不太顺，参考 numpy 官方的解释
random samples from a uniform distribution over [0, 1).

File: oneflow/python/test/ops/test_bce_with_logits_loss.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    input = np.random.random(size=input_shape).astype(np.float32)
```
Comment (Ldpe2G) [inline oneflow/python/test/ops/test_bce_with_logits_loss.py:36]: input 的范围是 [0,1]，没有负数部分，减个 0.5？或者用其他随机函数

File: oneflow/python/ops/nn_ops.py (hunk 5 above)
Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py:2967]: input = 1的时候也可以满足， pytorch也支持这种情况。 他默认将类别放在最后一维度，所以这里检查 pos_weight.shape[0] (因为他就只有一个维度） 是否等于 类别数目

Comment (Ldpe2G) [reply oneflow/python/ops/nn_ops.py:2967]: input 等于1维就是说只有一个样本？

**Review by Ldpe2G — APPROVED** [review]
